lib/std/special/compiler_rt/emutls.zig

```diff
@@ -0,0 +1,356 @@
+    }
+
+    /// deallocate the ObjectArray.
+    pub fn deinit(self: *ObjectArray) void {
+        const allocator = self.allocator;
+
+        // deallocated used objects in the array
+        for (self.slots) |_object, i| {
+            if (self.slots[i] != null) {
+                // XXX slot allocated using posix_memalign
+                std.c.free(self.slots[i]);
+                self.slots[i] = null;
+            }
+        }
+        allocator.free(self.slots);
+        self.* = undefined;
+        allocator.destroy(self);
+    }
+
+    /// resize the ObjectArray if needed.
+    pub fn ensureSize(self: *ObjectArray, new_size: usize) !void {
+        const old_size = self.slots.len;
+
+        if (old_size > new_size)
+            return;
+
+        // reallocate
+        self.slots = try self.allocator.reallocAtLeast(self.slots, new_size);
+
+        // init newly added slots
+        for (self.slots[old_size..]) |_object, i| {
+            self.slots[old_size + i] = null;
+        }
+    }
+
+    /// Retrieve the pointer at request index, using control to initialize it if needed.
+    pub inline fn get_pointer(self: *ObjectArray, index: usize, control: *emutls_control) !ObjectPointer {
+        if (self.slots[index] == null) {
+            // initialize the slot
+            const size = control.size;
+            const alignment = std.math.max(
+                @truncate(u29, control.alignment),
+                @alignOf(usize),
+            );
+
+            // XXX slot allocated using posix_memalign
+            var aligned_ptr: ?*c_void = undefined;
+            if (std.c.posix_memalign(&aligned_ptr, alignment, size) != 0) {
+                @panic("posix_memalign");
+            }
+            errdefer std.c.free(aligned_ptr);
+
+            if (control.default_value) |value| {
+                // default value: copy the content to newly allocated object.
+                @memcpy(@ptrCast([*]u8, aligned_ptr), @ptrCast([*]u8, value), size);
+            } else {
+                // no default: return zeroed memory.
+                @memset(@ptrCast([*]u8, aligned_ptr), 0, size);
+            }
+
+            self.slots[index] = @ptrCast(*c_void, aligned_ptr);
+        }
+
+        return self.slots[index].?;
+    }
+};
+
+// global stucture for ThreadStorage.
+// it provides thread-safety for on-demand storage of Thread Objects.
+const ThreadStorage = struct {
+    var key: std.c.pthread_key_t = undefined;
+    var init_once = std.once(ThreadStorage.init);
+
+    /// Return a per thread ObjectArray with at least the expected index.
+    pub inline fn get_array(allocator: *Allocator, index: usize) !*ObjectArray {
+        if (ThreadStorage.getspecific()) |array| {
+            // ensure the array is big enough for the wanted index
+            try array.ensureSize(index);
+            return array;
+        } else {
+            // a new array should contains at least 16 objects.
+            const size = std.math.max(16, index);
```

Comment:
So, if `index>16` you allocate only 16 slots and then leave `get_pointer` to read past `slots` end? `get_pointer` has no checks beside the ones inserted by the compiler in debug mode.

Reply:
if `index>16`, `std.math.max()` will return index value, and not 16.

the initial size to 16 is used to avoid too much reallocation at startup.

Reply:
Gah, my eyes kept reading `max` as `min`.
Yeah, 16 is enough for the common Zig program.